dot/state/epoch.go

```diff
@@ -397,26 +379,36 @@ func (s *EpochState) getConfigDataFromDatabase(epoch uint64) (*types.ConfigData,
 	return info, nil
 }
 
-// getConfigDataFromMemory retrieves the BABE config data for a given epoch that belongs to the header parameter
-func (s *EpochState) getConfigDataFromMemory(epoch uint64, header *types.Header) (*types.ConfigData, error) {
-	s.nextConfigDataLock.RLock()
-	defer s.nextConfigDataLock.RUnlock()
+func retrieveFromMemory[T types.NextEpochData | types.NextConfigData](nextEpochMap map[uint64]map[common.Hash]T,
+	es *EpochState, epoch uint64, header *types.Header) (*T, error) {
 
-	atEpoch, has := s.nextConfigData[epoch]
+	atEpoch, has := nextEpochMap[epoch]
 	if !has {
 		return nil, fmt.Errorf("%w: %d", ErrEpochNotInMemory, epoch)
 	}
 
 	headerHash := header.Hash()
-
 	for hash, value := range atEpoch {
-		isDescendant, err := s.blockState.IsDescendantOf(hash, headerHash)
+		isDescendant, err := es.blockState.IsDescendantOf(hash, headerHash)
+
+		// sometimes while moving to the next epoch is possible the header
+		// is not fully imported by the blocktree, in this case we will use
+		// its parent header which migth be already imported.
+		if errors.Is(err, blocktree.ErrEndNodeNotFound) {
+			parentHeader, err := es.blockState.GetHeader(header.ParentHash)
+			if err != nil {
+				return nil, fmt.Errorf("cannot get parent header: %w", err)
+			}
+
+			return retrieveFromMemory(nextEpochMap, es, epoch, parentHeader)
+		}
```

Comment:
what do you mean the header is not fully imported? a block can only be imported or not imported afaik

Reply:
The problem I notice is that at the block where we should change to the next epoch, I got the error `blocktree.ErrEndNodeNotFound` on function `chainProcessor.processBlockData`.

This method (`processBlockData`) calls the method `handleHeader` which calls the `babeVerifier.VerifyBlock(header)`, and the variable `header` contains the header value of the first block in the next epoch. The `VerifyBlock` method (`lib/babe/verify.go`) calls the method `getVerifierInfo` which calls the `epochState.GetEpochData` using the variables `header` and `epoch`. 

The `epochState.GetEpochData` did not find anything in the database for epoch 1 so we look up in the memory map using the `header` and that is when the problem happens, the block state is not aware of the `header` (first block in the next epoch) and return: `end node does not exist`. 

The fully error is:
```
2022-06-08T18:30:12+05:30 EROR block data processing for block with hash 0xd671d24ff29cf990010bfc98a54981e06079ad71f900f86881e30b186ea5d158 failed: could not verify block: failed to get verifier info for block 23: failed to get epoch data for epoch 1: failed to get epoch data from memory: cannot verify the ancestry: end node does not exist	chain_processor.go:L84	pkg=sync
```

Reply:
that makes sense, this seems like a decent fix, but maybe we should just pass the parent hash into this function instead, since the parent would always be imported already (i think). what do you think?